text/entities/0256-entities-data-model.md

```diff
@@ -0,0 +1,790 @@
+    <td><strong>Field</strong>
+    </td>
+    <td><strong>Type</strong>
+    </td>
+    <td><strong>Description</strong>
+    </td>
+   </tr>
+   <tr>
+    <td>Type
+    </td>
+    <td>string
+    </td>
+    <td>Defines the type of the entity. MUST not change during the
+lifetime of the entity. For example: "service" or "host". This field is
+required and MUST not be empty for valid entities.
+    </td>
+   </tr>
+   <tr>
+    <td>Id
+    </td>
+    <td>map&lt;string, attribute&gt;
+    </td>
+    <td>Attributes that identify the entity.
+<p>
+MUST not change during the lifetime of the entity. The Id must contain
+at least one attribute.
+<p>
+Follows OpenTelemetry <a
+href="https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute">common
+attribute definition</a>. SHOULD follow OpenTelemetry <a
+href="https://github.com/open-telemetry/semantic-conventions">semantic
+conventions</a> for attributes.
+    </td>
+   </tr>
+   <tr>
+    <td>Attributes
+    </td>
+    <td>map&lt;string, any&gt;
+    </td>
+    <td>Descriptive (non-identifying) attributes of the entity.
+<p>
+MAY change over the lifetime of the entity. MAY be empty. These
+attributes are not part of entity's identity.
+<p>
+Follows <a
+href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#type-any">any</a>
+value definition in the OpenTelemetry spec - it can be a scalar value,
+byte array, an array or map of values. Arbitrary deep nesting of values
+for arrays and maps is allowed.
+<p>
+SHOULD follow OpenTelemetry <a
+href="https://github.com/open-telemetry/semantic-conventions">semantic
+conventions</a> for attributes.
+    </td>
+   </tr>
+</table>
+
+### Minimally Sufficient Id
+
+Commonly, a number of attributes of an entity are readily available for the telemetry
+producer to compose an Id from. Of the available attributes the entity Id should
+include the minimal set of attributes that is sufficient for uniquely identifying
+that entity. No superfluous attributes should be included in the Id set. For example
+a Process on a host can be uniquely identified by (`process.pid`,`process.start_time`)
+attributes. Adding for example `process.executable.name` attribute to the Id is
+unnecessary and violates the Minimally Sufficient Id rule.
+
+### Examples of Entities
+
+_This section is non-normative and is present only for the purposes of demonstrating
+the data model._
+
+Here are examples of entities, the typical identifying attributes they
+have and some examples of non-identifying attributes that may be
+associated with the entity.
+
+<table>
+   <tr>
+    <td><strong>Entity</strong>
+    </td>
+    <td><strong>Entity Type</strong>
+    </td>
+    <td><strong>Identifying Attributes</strong>
+    </td>
+    <td><strong>Non-identifying Attributes</strong>
+    </td>
+   </tr>
+   <tr>
+    <td>Service
+    </td>
+    <td>"service"
+    </td>
+    <td>service.name (required)
```

Comment:
Does this mean that an entity type is not guaranteed to always have the same identifying attribute key set?

How will we distinguish between something with `service.name=foo` and `service.name=foo, service.instance.id=1`? Won't there be ambiguity when the entities are retrieved?

Reply:
>Won't there be ambiguity when the entities are retrieved?

Yes, there will be so, the answer is "don't do that". The guarantees will come from the producers of the entity information. It's their responsibility to produce valid, non-ambiguous Ids. Generally, we heavily rely on producers doing the right thing.

Also, this is just an example, it is not normative for this OTEP. The exact conventions for services and what is a required vs non-required attribute TBD, we have a TODO for that: https://github.com/orgs/open-telemetry/projects/85/views/1?pane=issue&itemId=63999527

Reply:
Do we expect producers will need to send "unset" values to try guarantee the uniformity of Id attributes? Does that mean we need a `nil` value for attributes?

Reply:
>Does that mean we need a nil value for attributes?

I had thought about that a while back and decided against it. It is a valid situation that different instances of entities of the same type have different sets of Id attributes. I don't think there is a need to include `nil` valued attributes just to ensure entities of the same type always have the same exact set of attributes.

If there is a use case where such uniformity is desired it can be achieved by read-time on-the-fly backfilling of missing attributes by `nil` values (and I am not sure there is a use case like that).

Reply:
I imagine the use-case is the error case where we want to degrade gracefully by reporting we know we should add an Id attribute but couldn't determine the value.

I ask about `nil` though, because the [common attributes](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute) call out that setting a `null` value is not allowed. If we do want to allow this possibility in the future we would need to use different attributes now or we would have to change the attributes specification.

This doesn't seem like a blocker, but I want to call it out that if we go with this design there will not be a way for a producer to always guarantee the uniformity of Ids with the common attributes.

Reply:
>I imagine the use-case is the error case where we want to degrade gracefully by reporting we know we should add an Id attribute but couldn't determine the value.

Good point. I added a TODO to think about this: https://github.com/orgs/open-telemetry/projects/85/views/1?pane=issue&itemId=64166000